include/mbedtls/ssl.h

```diff
@@ -2737,15 +2737,19 @@ const mbedtls_x509_crt *mbedtls_ssl_get_peer_cert( const mbedtls_ssl_context *ss
  * \brief          Save session in order to resume it later (client-side only)
  *                 Session data is copied to presented session structure.
  *
- * \warning        Currently, peer certificate is lost in the operation.
  *
  * \param ssl      SSL context
  * \param session  session context
  *
  * \return         0 if successful,
  *                 MBEDTLS_ERR_SSL_ALLOC_FAILED if memory allocation failed,
  *                 MBEDTLS_ERR_SSL_BAD_INPUT_DATA if used server-side or
- *                 arguments are otherwise invalid
+ *                 arguments are otherwise invalid.
+ *
+ * \note           Only the server certificate is copied, and not the chain
+ *                 but this is not a problem because the result of the chain
```

Comment:
Why are you mentioning this if it isn't a problem?

I think you wrote this paragraph from the point of view of the implementation. Please put yourself in the shoes of a user of the library. I've decided to use this function, or I'm considering using this function… What happens? If I access the server certificate, where does its `next` field point to? What memory do I need to free when I'm done with the copied object?

(Note: I deliberately reviewed without looking at the code of this function.)

Reply:
See https://github.com/ARMmbed/mbedtls/pull/927#issuecomment-302338235

Reply:
I agree that "but this is not a problem" is not clear and should be replaced with detailled instructions to the user, along the lines of: "Only the server certificate is copied, and not the full chain, so you should not attempt to validate the certificate again by calling `mbedtls_x509_crt_verify()` on it; instead, you can use the results from the verification in the original handshake by calling `mbedtls_ssl_get_verify_result()` after loading the session again into a new SSL context using `mbedtls_ssl_set_session()`."

While this is independent from the "but this is not a problem" point, I think it's also worth answering Gilles' question about what the user should free and when: the answer is you just need to call `mbedtls_ssl_session_free()` on the session object when you don't need it any more, and you can do this even if an SSL context that loaded this session using `mbedtls_ssl_set_session()` is still alive.